framework/gltf_loader.cpp

```diff
@@ -501,46 +501,59 @@ sg::Scene GLTFLoader::load_scene(int scene_index)
 	}
 
 	std::vector<std::unique_ptr<sg::Image>> image_components;
-	for (auto &fut : image_component_futures)
+
+	// Upload images to GPU. We do this in batches of 64MB of data to avoid needing
+	// double the amount of memory (all the images and all the corresponding buffers).
+	// This helps keep memory footprint lower which is helpful on smaller devices.
+	size_t image_index = 0;
+	while (image_index < image_count)
 	{
-		image_components.push_back(fut.get());
-	}
+		std::vector<core::Buffer> transient_buffers;
 
-	// Upload images to GPU
-	std::vector<core::Buffer> transient_buffers;
+		auto &command_buffer = device.request_command_buffer();
 
-	auto &command_buffer = device.request_command_buffer();
+		command_buffer.begin(VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT, 0);
 
-	command_buffer.begin(VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT, 0);
+		size_t batch_size = 0;
 
-	for (size_t image_index = 0; image_index < image_count; image_index++)
-	{
-		auto &image = image_components.at(image_index);
+		// Deal with 64MB of image data at a time to keep memory footprint low
+		while (image_index < image_count && batch_size < 64 * 1024 * 1024)
+		{
+			// Wait for this image to complete loading, then stage for upload
+			image_components.push_back(image_component_futures.at(image_index).get());
 
-		core::Buffer stage_buffer{device,
-		                          image->get_data().size(),
-		                          VK_BUFFER_USAGE_TRANSFER_SRC_BIT,
-		                          VMA_MEMORY_USAGE_CPU_ONLY};
+			auto &image = image_components.at(image_index);
 
-		stage_buffer.update(image->get_data());
+			core::Buffer stage_buffer{device,
+			                          image->get_data().size(),
+			                          VK_BUFFER_USAGE_TRANSFER_SRC_BIT,
+			                          VMA_MEMORY_USAGE_CPU_ONLY};
 
-		upload_image_to_gpu(command_buffer, stage_buffer, *image);
+			batch_size += image->get_data().size();
 
-		transient_buffers.push_back(std::move(stage_buffer));
-	}
+			stage_buffer.update(image->get_data());
 
-	command_buffer.end();
+			upload_image_to_gpu(command_buffer, stage_buffer, *image);
 
-	auto &queue = device.get_queue_by_flags(VK_QUEUE_GRAPHICS_BIT, 0);
+			transient_buffers.push_back(std::move(stage_buffer));
 
-	queue.submit(command_buffer, device.request_fence());
+			image_index++;
+		}
 
-	device.get_fence_pool().wait();
-	device.get_fence_pool().reset();
-	device.get_command_pool().reset_pool();
-	device.wait_idle();
+		command_buffer.end();
+
+		auto &queue = device.get_queue_by_flags(VK_QUEUE_GRAPHICS_BIT, 0);
+
+		queue.submit(command_buffer, device.request_fence());
 
-	transient_buffers.clear();
+		device.get_fence_pool().wait();
+		device.get_fence_pool().reset();
+		device.get_command_pool().reset_pool();
+		device.wait_idle();
```

Comment:
There's probably no need to wait_idle here.

Reply:
Again, not my code. Just rearranged slightly.